### PR TITLE
Fixed #25513 -- Extracted admin pagination to Paginator.get_elided_page_range().

### PR DESCRIPTION
**Pull request by pope1ni**
Supersedes #5866 & #10328 for ticket-25513.

Notes on changes:

- Pagination code in the admin is now 1-indexed. Backward incompatibility note added [as requested](https://github.com/django/django/pull/5866#discussion_r50706695). I think this will be less confusing in the long run, despite Carlton's [reservations](https://github.com/django/django/pull/5866#issuecomment-408768724). It is unlikely that a bookmark to a specific page will show consistent results over time anyway.

- ``DOT`` constant is now ``ELLIPSIS`` using ``'…'`` which makes it easier to use the correct character for display purposes and better fits with the new ``EllipsisPaginator`` class.

- Calculations have been corrected to avoid inserting ``ELLIPSIS`` to replace one number, e.g. ``[1, 2, ELLIPSIS, 4, 5, …]`` should be ``[1, 2, 3, 4, 5, …]``

- Documentation has been updated and moved to the correct location below the documentation for ``Paginator``.

Outstanding questions:

- [x] Should ``ELLIPSIS`` be moved to ``EllipsisPaginator.ELLIPSIS``? Then it can be easily documented and used in code that iterates over the ellipsized page range.

- [x] Should we look to [merge](https://github.com/django/django/pull/5866#issuecomment-173279531) this into the base ``Paginator`` class? We can't add it to ``page_range`` as that is a property - we need to pass the current page number - so it'd just be adding the new method instead of defining it on a separate class.

- [ ] ~~Should we perform any validation on the values provided to ``on_each_side`` and ``on_ends``?~~

- [x] "ellipsized" is a neologism, but the "ellipsed" used in a previous version of this doesn't quite feel right either. Any thoughts? _Changed to "elided"._

### CODE REVIEW

File: django/core/paginator.py [hunk 1]
```diff
@@ -7,6 +7,8 @@
 from django.utils.inspect import method_has_no_args
 from django.utils.translation import gettext_lazy as _
 
+ELLIPSIS = '…'
```
Comment (claudep) [inline django/core/paginator.py:10]: Should we mark it for translation (with a comment for translators)? I can imagine that some languages are using a different character. In doubt, we can also ask on gnome-i18n.

Comment (pope1ni) [reply django/core/paginator.py:10]: Seems reasonable.

**Comment by carltongibson** [comment]
> ... despite Carlton's reservations. 

Good. I'm glad that my initial response looking at this now was the same as last time. At least I'm consistent. 😀 — I shall try to convince myself that the behaviour change is OK here.

**Comment by carltongibson** [comment]
> "ellipsized" is a neologism, but the "ellipsed" used in a previous version of this doesn't quite feel right either. Any thoughts?

I think the verb is _to elide_, so _elided_, even though it's not identical.

**Comment by pope1ni** [comment]
> I think the verb is _to elide_, so _elided_, even though it's not identical?

So, yes, [ellipsed](https://www.thesaurus.com/browse/ellipse) is definitely not right. Thanks, [elided](https://www.thesaurus.com/browse/elide) fits much better. The only other options I can think of are "contracted" or "truncated". I'll change this.

> I shall try to convince myself that the behaviour change is OK here.

I think that the existing behaviour here is unexpected. Being a lazy fellow who likes to use the keyboard and avoid the mouse, I sometimes will edit the location bar and append `?p=5` if I want to jump to the fifth page rather than scroll to the bottom of the page and click a link. Also, if the page is in the elided range there is no button available anyway. I've always assumed this would take me to the fifth page and not the sixth.

The only minor concern I can see here is whether users are bookmarking certain pages, but I cannot see this being robust in any case as search results are affected significantly by data insertions, updates and deletions, as well as other factors such as filtering and sort order.

> * Should we look to [merge](https://github.com/django/django/pull/5866#issuecomment-173279531) this into the base `Paginator` class?

Thinking more about this, I'm +1 for doing this if you agree. Essentially we are only adding an additional method and developers can simply change `paginator.page_range` to `paginator.get_elided_page_range(number)` to benefit immediately without needing to change the class.

The main question is whether we add the `on_each_side` and `on_ends` arguments to the `__init__()` as we currently do or, since `get_elided_page_range()` must be called with the current page number anyway, whether we simply add them as keyword arguments there and keep everything self contained?

**Comment by carltongibson** [comment]
> Thinking more about this, I'm +1 for doing this if you agree.... 

OK, if you want to look at that, I won't object on principle. 🙂

> ...whether we simply add them as keyword arguments there and keep everything self contained?

This. Complicating `__init__()` on the base class for an edge behaviour is making everyone think about it for only a (small?) subset of uses. Optional kwargs on `get_elided_page_range()`, as you say, keeps everything self contained, so is better I think.

**Comment by pope1ni** [comment]
Ok. So I've separated out two commits to fix the admin pagination to be 1-indexed and to avoid elision of a single page number.

I've also condensed everything based on my comments above and expanded the tests significantly to cover all edge cases. It has turned out rather nicely. I think this is mostly there short of some tweaks to documentation and a few extra test assertions related to ensuring an iterator/generator is returned and the appropriate exceptions are raised.

File: docs/ref/paginator.txt [hunk 2]
```diff
@@ -78,9 +78,33 @@ Methods
     Returns a :class:`Page` object with the given 1-based index. Raises
     :exc:`InvalidPage` if the given page number doesn't exist.
 
+.. method:: Paginator.get_elided_page_range(number, *, on_each_side=3, on_ends=2)
+
```
Comment (felixxm) [inline docs/ref/paginator.txt:82]: `versionadded`

File: docs/ref/paginator.txt [hunk 3]
```diff
@@ -78,9 +78,33 @@ Methods
...
+    if the current page is 10 and there are 50 pages, the page range will be
+    ``[1, 2, '…', 7, 8, 9, 10, 11, 12, 13, '…', 49, 50]``. This will result in
+    pages 4, 5, and 6 to the left of and 8, 9, and 10 to the right of the
+    current page as well as pages 1 and 2 at the start and 49 and 50 at the
+    end.
+
 Attributes
 ----------
 
+.. attribute:: Paginator.ELLIPSIS
```
Comment (felixxm) [inline docs/ref/paginator.txt:103]: `versionadded`

File: tests/pagination/tests.py [hunk 4]
```diff
@@ -304,6 +305,120 @@ def test_paginator_iteration(self):
             with self.subTest(page=page):
                 self.assertEqual(expected, list(next(page_iterator)))
 
+    def test_get_elided_page_range(self):
+        # Check that Paginator.validate_number() is called:
```
Comment (felixxm) [inline tests/pagination/tests.py:309]: Try to avoid `Check that` and `Check`.

File: django/contrib/admin/templatetags/admin_list.py [hunk 5]
```diff
@@ -24,64 +24,31 @@
...
-
 
 @register.simple_tag
 def paginator_number(cl, i):
     """
     Generate an individual page index link in a paginated list.
     """
-    if i == DOT:
-        return '… '
+    if i == cl.paginator.ELLIPSIS:
```
Comment (felixxm) [inline django/contrib/admin/templatetags/admin_list.py:33]: Should we call it "ellipsis"? It's basically an option to customize a separator and do not use ellipsis :thinking:

File: docs/releases/3.2.txt [hunk 6]
```diff
@@ -345,6 +345,12 @@ Validators
 * The :class:`.ValidationError` equality operator now ignores ``messages`` and
   ``params`` ordering.
 
+Miscellaneous
```
Comment (felixxm) [inline docs/releases/3.2.txt:348]: I would add `Pagination` like in the previous release notes.

File: docs/releases/3.2.txt [hunk 7]
```diff
@@ -345,6 +345,12 @@ Validators
...
+~~~~~~~~~~~~~
+
+* :meth:`django.core.paginator.Paginator.get_elided_page_range` has been
```
Comment (felixxm) [inline docs/releases/3.2.txt:351]: Maybe:
```
The new :meth:`django.core.paginator.Paginator.get_elided_page_range` allows
customizing ... on ``Paginator`` subclasses ...
```

**Comment by carltongibson** [comment]
Hey @pope1ni: 

> ... a few extra test assertions related to ensuring an iterator/generator is returned and the appropriate exceptions are raised.

If you just want to add those we can push this in. (General outline is 👍 — Thanks for your work!)

File: django/contrib/admin/templatetags/admin_list.py (hunk 5 above)
Comment (pope1ni) [reply django/contrib/admin/templatetags/admin_list.py:33]: Yeah. So I changed it to `ELLIPSIS` from `DOT` as the intent is to make it easy to use by having a standard character for the elided sections of the page range that can just be pushed out for display purposes. It has also been made translatable for that reason, although another language may not use an ellipsis.

I'm struggling to think of a better name for this. Perhaps `ELIDE`, `SKIP`, `OMIT`, `BREAK`, or `CUT`?

File: docs/releases/3.2.txt (hunk 7 above)
Comment (pope1ni) [reply docs/releases/3.2.txt:351]: I don't think we really need to mention subclasses here. Ultimately I expect this to be flexible enough for nearly all use cases as the values of `on_ends` and `on_each_side` can be tweaked to handle almost every combination.

File: tests/pagination/tests.py [hunk 8]
```diff
@@ -304,6 +305,120 @@ def test_paginator_iteration(self):
...
+        paginator = Paginator([1, 2, 3], 2)
+        with unittest.mock.patch.object(paginator, 'validate_number') as mock:
+            mock.assert_not_called()
+            list(paginator.get_elided_page_range(2))
+            mock.assert_called_with(2)
```
Comment (pope1ni) [inline tests/pagination/tests.py:314]: @carltongibson Regarding my comment about testing exceptions, I've just remembered that I added this to check that `Paginator.validate_number()` is called which raises those exceptions for invalid page numbers. That is tested by `PaginationTests.test_invalid_page_number()` already.

Comment (carltongibson) [reply tests/pagination/tests.py:314]: Okay, super, thanks for the clarification @pope1ni. I shall take another look tomorrow and hopefully that's job done! Good work, as ever. 😉

File: django/contrib/admin/templatetags/admin_list.py (hunk 5 above)
Comment (carltongibson) [reply django/contrib/admin/templatetags/admin_list.py:33]: Let's go with "ellipsis": it's primary meaning is _the omision_, for which folks _could_ use any symbol. It's secondary meaning is the three dots `…` that we actually use.

**Pull request merged**